pgrx/src/nullable.rs

```diff
@@ -0,0 +1,523 @@
+//! Traits and implementations of functionality relating to Postgres null
+//! and nullable types, particularly in relation to arrays and other container
+//! types which implement nullable (empty slot) behavior in a
+//! "structure-of-arrays" manner
+
+use std::{fmt::Debug, iter::Enumerate, marker::PhantomData};
+
+use bitvec::slice::BitSlice;
+
+use crate::Datum;
+
+pub enum Nullable<T> {
+    Valid(T),
+    Null,
+}
+
+impl<T> Debug for Nullable<T>
+where
+    T: Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Valid(value) => f.debug_tuple("Valid").field(value).finish(),
+            Self::Null => write!(f, "Null"),
+        }
+    }
+}
+
+impl<T> Clone for Nullable<T>
+where
+    T: Clone,
+{
+    fn clone(&self) -> Self {
+        match self {
+            Self::Valid(arg0) => Self::Valid(arg0.clone()),
+            Self::Null => Self::Null,
+        }
+    }
+}
+
+impl<T> PartialEq for Nullable<T>
+where
+    T: Eq,
+{
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Valid(l_value), Self::Valid(r_value)) => l_value == r_value,
+            _ => core::mem::discriminant(self) == core::mem::discriminant(other),
+        }
+    }
+}
+
+impl<T> Eq for Nullable<T> where T: PartialEq + Eq {}
+
+impl<T> Nullable<T> {
+    pub fn into_option(self) -> Option<T> {
+        match self {
+            Nullable::Valid(val) => Some(val),
+            Nullable::Null => None,
+        }
+    }
+    pub fn as_option<'a>(&'a self) -> Option<&'a T> {
+        match self {
+            Nullable::Valid(val) => Some(val),
+            Nullable::Null => None,
+        }
+    }
+    pub fn as_option_mut<'a>(&'a mut self) -> Option<&'a T> {
+        match self {
+            Nullable::Valid(val) => Some(val),
+            Nullable::Null => None,
+        }
+    }
+    pub fn is_valid(&self) -> bool {
+        match self {
+            Nullable::Valid(_) => true,
+            Nullable::Null => false,
+        }
+    }
+    pub fn is_null(&self) -> bool {
+        !self.is_valid()
+    }
+    pub fn unwrap(self) -> T {
+        self.into_option().unwrap()
+    }
+    pub fn expect(self, msg: &'static str) -> T {
+        self.into_option().expect(msg)
+    }
```

Comment:
I wonder if we want a different signature than std has for `expect`... I often find myself wishing it would take format-args, somehow...

Reply:
My worry would be people might be surprised by the different signature. However, something `&'static str` satisfies, but which is more flexible, might work out well - `T: Into<Cow<'a, str>>` perhaps? The standard library has from/into implementations from `&'a str` to `Cow<'a, str>`. I'll give that a shot.

Reply:
Implemented in 1a3d5719b77c233fd96440597e3aa43d794847d4